File: Talleres/Taller 2/code/BinaryInversedDyV.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Convert an integer number into binary list
std::string DecimalToBinaryDyV(int n)
{
    string binary = "";
    int i = 0;
    while (n > 0)
    {
        char c = n % 2 + '0';
        binary += c;
        n = n / 2;
        i++;
    }
    return binary;
}

// Invert a chain given a list begin and end
string InvertChainDyV(string S, int l, int r)
{
    if (l == r)
    {
        string aux = "";
        aux += S[l];
        return aux;
    }

    int q = (l + r) / 2;
    string left = InvertChainDyV(S, l, q);
    string right = InvertChainDyV(S, q + 1, r);
    string result = "";

    for (int i = 0; i < right.size(); i++)
    {
        result += right[i];
    }
    for (int i = 0; i < left.size(); i++)
    {
        result += left[i];
    }
    return result;
}

int BinaryToDecimalDyV(string S)
{
    int decimal = 0;
    int i = 0;
    string aux = InvertChainDyV(S, 0, S.size() - 1);
    for (int i = 0; i < aux.size(); i++)
    {
        int c = aux[i] - '0';
        int p = c * pow(2, aux.size() - i - 1);
        decimal += p;
    }
    return decimal;
}

File: Talleres/Taller 2/code/python/BinaryInversedDyV.py
from math import pow
#Convert an integer number into binary list
def DecimalToBinary(n):
    binary = []
    while n != 0:
        binary.insert(0, str(n % 2))
        n //= 2
    return (binary)

#Invert a chain given a list begin and end
def InvertChain(S,l,r):
    if l==r:
        return S[l]

    q = (l+r)//2
    left = InvertChain(S,l,q)
    right= InvertChain(S,q+1,r)
    result=[]

    #SWAP
    for i in range(len(right)):
        result.append(right[i])
    for i in range(len(left)):
        result.append(left[i])
    return result

#Convert a list of binary into an integer
def BinaryToDecimal(numero_binario):
    decimal = 0
    i=0
    inverted = InvertChain(numero_binario,0,len(numero_binario)-1)
    for i in range(len(inverted)):
        p = int(pow(int(inverted[i])*2,i))
        decimal = decimal + p
    return decimal

print("Binary Inversed DyV")
n = 345
binary= DecimalToBinary(n)
print("Original:", n, binary)
binarioInverso = InvertChain(binary,0,len(binary)-1)
decimal= BinaryToDecimal(binarioInverso)
print("Inverso:", decimal, binarioInverso)